### PR TITLE
Application base

### PR DESCRIPTION
**Pull request by jhryniw**
The base application with a default LoginActivity.

Let me know if you want to create a `develop` branch and I can cancel this PR.

### CODE REVIEW

File: app/src/main/java/cmput301f17t13/com/catisadog/LoginActivity.java [hunk 1]
```diff
@@ -0,0 +1,350 @@
...
+    private EditText mPasswordView;
+    private View mProgressView;
+    private View mLoginFormView;
+
+    @Override
+    protected void onCreate(Bundle savedInstanceState) {
+        super.onCreate(savedInstanceState);
+        setContentView(R.layout.activity_login);
+        // Set up the login form.
+        mEmailView = (AutoCompleteTextView) findViewById(R.id.email);
```
Comment (linster) [inline app/src/main/java/cmput301f17t13/com/catisadog/LoginActivity.java]: Check out Butterknife for avoiding lots of findViewById calls :)

File: app/src/main/java/cmput301f17t13/com/catisadog/models/FollowRequest.java [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+
+import org.joda.time.DateTime;
+
+import java.util.UUID;
+
+/**
+ * Created by James on 2017-10-20.
+ */
+
+public class FollowRequest {
```
Comment (linster) [inline app/src/main/java/cmput301f17t13/com/catisadog/models/FollowRequest.java]: Classes like these are a great way to get into Kotlin. 

https://kotlinlang.org/docs/reference/data-classes.html

File: app/src/main/java/cmput301f17t13/com/catisadog/LoginActivity.java [hunk 3]
```diff
@@ -0,0 +1,350 @@
...
+            // perform the user login attempt.
+            showProgress(true);
+            mAuthTask = new UserLoginTask(email, password);
+            mAuthTask.execute((Void) null);
+        }
+    }
+
+    private boolean isEmailValid(String email) {
+        //TODO: Replace this with your own logic
+        return email.contains("@");
```
Comment (linster) [inline app/src/main/java/cmput301f17t13/com/catisadog/LoginActivity.java]: Check out https://developer.android.com/reference/android/util/Patterns.html#EMAIL_ADDRESS

File: app/src/main/java/cmput301f17t13/com/catisadog/models/FollowRequest.java (hunk 2 above)
Comment (jhryniw) [reply app/src/main/java/cmput301f17t13/com/catisadog/models/FollowRequest.java]: As much as I'd like to get into Kotlin, the class demands Vanilla java

File: app/src/main/java/cmput301f17t13/com/catisadog/LoginActivity.java (hunk 1 above)
Comment (jhryniw) [reply app/src/main/java/cmput301f17t13/com/catisadog/LoginActivity.java]: Looks promising! Note that I just put in the default LoginActivity as a starting point and major changes will be made.

File: app/src/main/java/cmput301f17t13/com/catisadog/LoginActivity.java (hunk 3 above)
Comment (jhryniw) [reply app/src/main/java/cmput301f17t13/com/catisadog/LoginActivity.java]: We won't be using emails in the final design.

File: app/src/main/java/cmput301f17t13/com/catisadog/models/FollowRequest.java (hunk 2 above)
Comment (jhryniw) [reply app/src/main/java/cmput301f17t13/com/catisadog/models/FollowRequest.java]: Instead, we will be using https://github.com/google/auto/tree/master/value

Comment (linster) [reply app/src/main/java/cmput301f17t13/com/catisadog/models/FollowRequest.java]: Aww man, sorry to hear about the Kotlin Embargo :|

File: app/src/androidTest/java/cmput301f17t13/com/catisadog/ExampleInstrumentedTest.java [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+
+/**
+ * Instrumentation test, which will execute on an Android device.
+ *
+ * @see <a href="http://d.android.com/tools/testing">Testing documentation</a>
+ */
+@RunWith(AndroidJUnit4.class)
+public class ExampleInstrumentedTest {
+    @Test
+    public void useAppContext() throws Exception {
```
Comment (linster) [inline app/src/androidTest/java/cmput301f17t13/com/catisadog/ExampleInstrumentedTest.java]: If you're writing tests that use Android-isms (context and the like), check out Robolectric. With Robolectric you can write tests that use android components without needing to fully instrument the test.

**Comment by lelandjansen** [comment]
> Let me know if you want to create a develop branch and I can cancel this PR.
My preference is to make `master` the "bleeding edge" meaning we always branch, rebase, and squash on top of master (so no `dev` branch). We can mark certain milestones with "release" tags.

File: app/src/androidTest/java/cmput301f17t13/com/catisadog/ExampleInstrumentedTest.java (hunk 4 above)
Comment (lelandjansen) [reply app/src/androidTest/java/cmput301f17t13/com/catisadog/ExampleInstrumentedTest.java]: Link: http://robolectric.org

Interesting idea, @linster. Thanks for the recommendation :)

**Comment by lelandjansen** [comment]
This PR is **WAY** too long to effectively review. I understand initial setup PRs tend to be longer with a bunch of setup/image/resources code but 60 file changes and >1800 loc is excessive. Please break this up into individual PRs.

File: app/src/main/java/cmput301f17t13/com/catisadog/activities/OtherUsersActivity.java [hunk 5]
```diff
@@ -0,0 +1,11 @@
+package cmput301f17t13.com.catisadog.activities;
+
+import android.app.Activity;
+
+/**
+ * Created by James on 2017-10-20.
```
Comment (lelandjansen) [inline app/src/main/java/cmput301f17t13/com/catisadog/activities/OtherUsersActivity.java]: Ditto.

File: app/src/main/java/cmput301f17t13/com/catisadog/fragments/summary/AddHabitEventFragment.java [hunk 6]
```diff
@@ -0,0 +1,10 @@
+package cmput301f17t13.com.catisadog.fragments.summary;
+
+import android.support.v4.app.FragmentActivity;
+
+/**
+ * Created by James on 2017-10-20.
```
Comment (lelandjansen) [inline app/src/main/java/cmput301f17t13/com/catisadog/fragments/summary/AddHabitEventFragment.java]: Ditto.

File: app/src/main/java/cmput301f17t13/com/catisadog/fragments/summary/AddHabitFragment.java [hunk 7]
```diff
@@ -0,0 +1,10 @@
+package cmput301f17t13.com.catisadog.fragments.summary;
+
+import android.support.v4.app.FragmentActivity;
+
+/**
+ * Created by James on 2017-10-20.
```
Comment (lelandjansen) [inline app/src/main/java/cmput301f17t13/com/catisadog/fragments/summary/AddHabitFragment.java]: Ditto.

File: app/src/main/java/cmput301f17t13/com/catisadog/fragments/summary/FollowingHabitsFragment.java [hunk 8]
```diff
@@ -0,0 +1,10 @@
+package cmput301f17t13.com.catisadog.fragments.summary;
+
+import android.support.v4.app.FragmentActivity;
+
+/**
+ * Created by James on 2017-10-20.
```
Comment (lelandjansen) [inline app/src/main/java/cmput301f17t13/com/catisadog/fragments/summary/FollowingHabitsFragment.java]: Ditto.

File: app/src/main/java/cmput301f17t13/com/catisadog/fragments/summary/MyHabitsFragment.java [hunk 9]
```diff
@@ -0,0 +1,10 @@
+package cmput301f17t13.com.catisadog.fragments.summary;
+
+import android.support.v4.app.FragmentActivity;
+
+/**
+ * Created by James on 2017-10-20.
```
Comment (lelandjansen) [inline app/src/main/java/cmput301f17t13/com/catisadog/fragments/summary/MyHabitsFragment.java]: Ditto.

File: app/src/main/java/cmput301f17t13/com/catisadog/fragments/summary/TodoHabitsFragment.java [hunk 10]
```diff
@@ -0,0 +1,11 @@
+package cmput301f17t13.com.catisadog.fragments.summary;
+
+import android.support.v4.app.FragmentActivity;
+
+/**
+ * Created by James on 2017-10-20.
```
Comment (lelandjansen) [inline app/src/main/java/cmput301f17t13/com/catisadog/fragments/summary/TodoHabitsFragment.java]: Ditto.

File: app/src/main/java/cmput301f17t13/com/catisadog/fragments/summary/ViewHabitFragment.java [hunk 11]
```diff
@@ -0,0 +1,10 @@
+package cmput301f17t13.com.catisadog.fragments.summary;
+
+import android.support.v4.app.FragmentActivity;
+
+/**
+ * Created by James on 2017-10-20.
```
Comment (lelandjansen) [inline app/src/main/java/cmput301f17t13/com/catisadog/fragments/summary/ViewHabitFragment.java]: Ditto.

File: app/src/main/java/cmput301f17t13/com/catisadog/models/Habit.java [hunk 12]
```diff
@@ -0,0 +1,82 @@
...
+
+
+import org.joda.time.DateTime;
+
+import java.util.Collection;
+import java.util.HashSet;
+import java.util.List;
+
+/**
+ * Created by James on 2017-10-20.
```
Comment (lelandjansen) [inline app/src/main/java/cmput301f17t13/com/catisadog/models/Habit.java]: Ditto.

File: app/src/main/java/cmput301f17t13/com/catisadog/activities/LoginActivity.java [hunk 13]
```diff
@@ -0,0 +1,355 @@
...
+public class LoginActivity extends AppCompatActivity implements LoaderCallbacks<Cursor> {
+
+    /**
+     * Id to identity READ_CONTACTS permission request.
+     */
+    private static final int REQUEST_READ_CONTACTS = 0;
+
+    /**
+     * A dummy authentication store containing known user names and passwords.
+     * TODO: remove after connecting to a real authentication system.
```
Comment (lelandjansen) [inline app/src/main/java/cmput301f17t13/com/catisadog/activities/LoginActivity.java]: #16
Whenever making a `TODO` item in the code, please create a GitHub issue and reference it. For example:
```
TODO(#16): Create issue for this TODO and reference the issue #
```

File: app/src/main/java/cmput301f17t13/com/catisadog/activities/LoginActivity.java [hunk 14]
```diff
@@ -0,0 +1,355 @@
...
+            // Show a progress spinner, and kick off a background task to
+            // perform the user login attempt.
+            showProgress(true);
+            mAuthTask = new UserLoginTask(email, password);
+            mAuthTask.execute((Void) null);
+        }
+    }
+
+    private boolean isEmailValid(String email) {
+        //TODO: Replace this with your own logic
```
Comment (lelandjansen) [inline app/src/main/java/cmput301f17t13/com/catisadog/activities/LoginActivity.java]: #16

File: app/src/main/java/cmput301f17t13/com/catisadog/activities/LoginActivity.java [hunk 15]
```diff
@@ -0,0 +1,355 @@
...
+        return email.contains("@");
+    }
+
+    private boolean isPasswordValid(String password) {
+        //TODO: Replace this with your own logic
```
Comment (lelandjansen) [inline app/src/main/java/cmput301f17t13/com/catisadog/activities/LoginActivity.java]: #16

File: app/src/main/java/cmput301f17t13/com/catisadog/activities/LoginActivity.java [hunk 16]
```diff
@@ -0,0 +1,355 @@
...
+        private final String mPassword;
+
+        UserLoginTask(String email, String password) {
+            mEmail = email;
+            mPassword = password;
+        }
+
+        @Override
+        protected Boolean doInBackground(Void... params) {
+            // TODO: attempt authentication against a network service.
```
Comment (lelandjansen) [inline app/src/main/java/cmput301f17t13/com/catisadog/activities/LoginActivity.java]: #16

File: app/src/main/java/cmput301f17t13/com/catisadog/activities/LoginActivity.java [hunk 17]
```diff
@@ -0,0 +1,355 @@
...
+
+            for (String credential : DUMMY_CREDENTIALS) {
+                String[] pieces = credential.split(":");
+                if (pieces[0].equals(mEmail)) {
+                    // Account exists, return true if the password matches.
+                    return pieces[1].equals(mPassword);
+                }
+            }
+
+            // TODO: register the new account here.
```
Comment (lelandjansen) [inline app/src/main/java/cmput301f17t13/com/catisadog/activities/LoginActivity.java]: #16

File: app/src/main/java/cmput301f17t13/com/catisadog/models/Habit.java [hunk 18]
```diff
@@ -0,0 +1,82 @@
...
+    private String title;
+    private String reason;
+
+    private DateTime startDate;
+    private HashSet<Integer> schedule;
+    private HabitStatus status;
+
+    String id;
+
+    //TODO: How to handle completion metrics
```
Comment (lelandjansen) [inline app/src/main/java/cmput301f17t13/com/catisadog/models/Habit.java]: #16

File: app/src/main/java/cmput301f17t13/com/catisadog/models/HabitEvent.java [hunk 19]
```diff
@@ -0,0 +1,76 @@
+package cmput301f17t13.com.catisadog.models;
+
+import org.joda.time.DateTime;
+
+import cmput301f17t13.com.catisadog.models.user.User;
+
+/**
+ * Created by Kevin on 2017-10-20.
```
Comment (lelandjansen) [inline app/src/main/java/cmput301f17t13/com/catisadog/models/HabitEvent.java]: Ditto.

File: app/src/main/java/cmput301f17t13/com/catisadog/models/HabitStatus.java [hunk 20]
```diff
@@ -0,0 +1,30 @@
+package cmput301f17t13.com.catisadog.models;
+
+import android.graphics.drawable.Drawable;
+
+/**
+ * Created by James on 2017-10-20.
```
Comment (lelandjansen) [inline app/src/main/java/cmput301f17t13/com/catisadog/models/HabitStatus.java]: Ditto.

File: app/src/main/java/cmput301f17t13/com/catisadog/models/Location.java [hunk 21]
```diff
@@ -0,0 +1,24 @@
+package cmput301f17t13.com.catisadog.models;
+
+/**
+ * Created blng Kevin on 2017-10-20.
```
Comment (lelandjansen) [inline app/src/main/java/cmput301f17t13/com/catisadog/models/Location.java]: Ditto.

File: app/src/main/java/cmput301f17t13/com/catisadog/models/Schedulable.java [hunk 22]
```diff
@@ -0,0 +1,21 @@
+package cmput301f17t13.com.catisadog.models;
+
+import org.joda.time.DateTime;
+
+import java.util.Collection;
+
+/**
+ * Created by James on 2017-10-20.
```
Comment (lelandjansen) [inline app/src/main/java/cmput301f17t13/com/catisadog/models/Schedulable.java]: Ditto.

File: app/src/main/java/cmput301f17t13/com/catisadog/models/data/Datastore.java [hunk 23]
```diff
@@ -0,0 +1,40 @@
...
+import java.util.HashSet;
+
+import cmput301f17t13.com.catisadog.models.user.FollowRequest;
+import cmput301f17t13.com.catisadog.models.Habit;
+import cmput301f17t13.com.catisadog.models.HabitEvent;
+import cmput301f17t13.com.catisadog.models.HabitStatus;
+import cmput301f17t13.com.catisadog.models.user.User;
+
+/**
+ * Created by Kevin on 2017-10-20.
```
Comment (lelandjansen) [inline app/src/main/java/cmput301f17t13/com/catisadog/models/data/Datastore.java]: Ditto.

File: app/src/main/java/cmput301f17t13/com/catisadog/models/data/FirebaseDatastore.java [hunk 24]
```diff
@@ -0,0 +1,103 @@
...
+import java.util.HashSet;
+
+import cmput301f17t13.com.catisadog.models.user.FollowRequest;
+import cmput301f17t13.com.catisadog.models.Habit;
+import cmput301f17t13.com.catisadog.models.HabitEvent;
+import cmput301f17t13.com.catisadog.models.HabitStatus;
+import cmput301f17t13.com.catisadog.models.user.User;
+
+/**
+ * Created by Kevin on 2017-10-20.
```
Comment (lelandjansen) [inline app/src/main/java/cmput301f17t13/com/catisadog/models/data/FirebaseDatastore.java]: Ditto.

File: app/src/main/java/cmput301f17t13/com/catisadog/models/data/GsonDatastore.java [hunk 25]
```diff
@@ -0,0 +1,103 @@
...
+import java.util.HashSet;
+
+import cmput301f17t13.com.catisadog.models.user.FollowRequest;
+import cmput301f17t13.com.catisadog.models.Habit;
+import cmput301f17t13.com.catisadog.models.HabitEvent;
+import cmput301f17t13.com.catisadog.models.HabitStatus;
+import cmput301f17t13.com.catisadog.models.user.User;
+
+/**
+ * Created by Kevin on 2017-10-20.
```
Comment (lelandjansen) [inline app/src/main/java/cmput301f17t13/com/catisadog/models/data/GsonDatastore.java]: Ditto.

File: app/src/main/java/cmput301f17t13/com/catisadog/models/user/CurrentUser.java [hunk 26]
```diff
@@ -0,0 +1,27 @@
+package cmput301f17t13.com.catisadog.models.user;
+
+import java.util.Collection;
+
+/**
+ * Created by Kevin on 2017-10-20.
```
Comment (lelandjansen) [inline app/src/main/java/cmput301f17t13/com/catisadog/models/user/CurrentUser.java]: Ditto.

File: app/src/main/java/cmput301f17t13/com/catisadog/models/user/FollowRequest.java [hunk 27]
```diff
@@ -0,0 +1,43 @@
+package cmput301f17t13.com.catisadog.models.user;
+
+import org.joda.time.DateTime;
+
+/**
+ * Created by James on 2017-10-20.
```
Comment (lelandjansen) [inline app/src/main/java/cmput301f17t13/com/catisadog/models/user/FollowRequest.java]: Ditto.

File: app/src/main/java/cmput301f17t13/com/catisadog/models/user/User.java [hunk 28]
```diff
@@ -0,0 +1,50 @@
+package cmput301f17t13.com.catisadog.models.user;
+
+import java.util.Collection;
+import java.util.List;
+
+/**
+ * Created by James on 2017-10-20.
```
Comment (lelandjansen) [inline app/src/main/java/cmput301f17t13/com/catisadog/models/user/User.java]: Ditto.

File: app/src/main/java/cmput301f17t13/com/catisadog/models/user/User.java [hunk 29]
```diff
@@ -0,0 +1,50 @@
...
+    public User(String username) {
+        this.username = username;
+    }
+
+    public String getUsername() {
+        return username;
+    }
+
+    public void follow(User otherUser) {
+        // Sends a follow request
```
Comment (lelandjansen) [inline app/src/main/java/cmput301f17t13/com/catisadog/models/user/User.java]: Should this be a TODO? #16

File: app/src/main/java/cmput301f17t13/com/catisadog/utils/LocationHelper.java [hunk 30]
```diff
@@ -0,0 +1,14 @@
+package cmput301f17t13.com.catisadog.utils;
+
+import cmput301f17t13.com.catisadog.models.Location;
+
+/**
+ * Created by Kevin on 2017-10-20.
```
Comment (lelandjansen) [inline app/src/main/java/cmput301f17t13/com/catisadog/utils/LocationHelper.java]: Ditto.

**Comment by lelandjansen** [comment]
Should `gradle/wrapper/gradle-wrapper.jar` be `.gitignore`'d?

File: app/src/main/java/cmput301f17t13/com/catisadog/activities/LoginActivity.java [hunk 31]
```diff
@@ -0,0 +1,355 @@
...
+
+import android.animation.Animator;
+import android.animation.AnimatorListenerAdapter;
+import android.annotation.TargetApi;
+import android.content.pm.PackageManager;
+import android.support.annotation.NonNull;
+import android.support.design.widget.Snackbar;
+import android.support.v7.app.AppCompatActivity;
+import android.app.LoaderManager.LoaderCallbacks;
+
```
Comment (lelandjansen) [inline app/src/main/java/cmput301f17t13/com/catisadog/activities/LoginActivity.java]: Why is there blank space between these imports (and others)?

File: app/src/main/java/cmput301f17t13/com/catisadog/activities/LoginActivity.java (hunk 13 above)
Comment (jhryniw) [reply app/src/main/java/cmput301f17t13/com/catisadog/activities/LoginActivity.java]: I didn't write any code in this file. I dumped in the default Android Studio Login Activity... we can delete it to a vanilla version if you want.

File: app/src/main/java/cmput301f17t13/com/catisadog/activities/LoginActivity.java (hunk 31 above)
Comment (jhryniw) [reply app/src/main/java/cmput301f17t13/com/catisadog/activities/LoginActivity.java]: Not sure why here (didn't make this file), but Android studio generally splits them by package highest order of package.

File: app/src/main/java/cmput301f17t13/com/catisadog/activities/LoginActivity.java (hunk 13 above)
Comment (lelandjansen) [reply app/src/main/java/cmput301f17t13/com/catisadog/activities/LoginActivity.java]: Up to you. If you keep it in please reference an issue (possibly related to a larger goal such as authentication)
.

File: app/src/main/java/cmput301f17t13/com/catisadog/activities/LoginActivity.java (hunk 31 above)
Comment (lelandjansen) [reply app/src/main/java/cmput301f17t13/com/catisadog/activities/LoginActivity.java]: Okay. Let's stick with Android Studio's convention.

File: app/src/main/java/cmput301f17t13/com/catisadog/models/user/User.java (hunk 29 above)
Comment (jhryniw) [reply app/src/main/java/cmput301f17t13/com/catisadog/models/user/User.java]: Everything is a TODO haha

**Comment by lelandjansen** [comment]
@jhryniw Thanks for the TODO references

File: app/src/androidTest/java/cmput301f17t13/com/catisadog/ExampleInstrumentedTest.java (hunk 4 above)
Comment (lelandjansen) [reply app/src/androidTest/java/cmput301f17t13/com/catisadog/ExampleInstrumentedTest.java]: I looked at the rubric and unfortunately we're forced to JUnit.

**Pull request merged**